### PR TITLE
Add customized result index in data source etc

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This PR
- Introduce `spark.flint.datasource.name` parameter for data source specification.
- Enhance data source creation to allow custom result indices; fallback to default if unavailable.
- Include error details in the async result response, sourced from the result index.
- Migrate to `org.apache.spark.sql.FlintJob` following updates in OpenSearch-Spark.
- Populate query status from result index over EMR-S job status to handle edge cases where jobs may succeed, but queries or mappings fail.

Testing done:
1. manual testing including if with or without custom result index async query still works
2. added new unit tests

 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java [hunk 1]
```diff
@@ -8,13 +8,22 @@
 public class SparkConstants {
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java:8]: could add the change in https://github.com/opensearch-project/sql/pull/2216/files#diff-6eca1b0df87741e862bd1455b8a39e503489990984f8d790ebf7de7016986b86

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 2]
```diff
@@ -117,14 +144,16 @@ private DispatchQueryResponse handleIndexQuery(
                 .build()
                 .toString(),
             tags,
-            indexDetails.getAutoRefresh());
+            indexDetails.getAutoRefresh(),
+            dataSourceMetadata.getResultIndex());
     String jobId = emrServerlessClient.startJobRun(startJobRequest);
-    return new DispatchQueryResponse(jobId, false);
+    return new DispatchQueryResponse(jobId, false, dataSourceMetadata.getResultIndex());
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:150]: what if resultIndex is null?

File: datasources/src/main/resources/datasources-index-mapping.yml [hunk 3]
```diff
@@ -14,4 +14,6 @@ properties:
       keyword:
         type: keyword
   connector:
+    type: keyword
+  resultIndex:
```
Comment (vamsi-amazon) [inline datasources/src/main/resources/datasources-index-mapping.yml:18]: Do we need indexing on this?

File: spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java (hunk 1 above)
Comment (kaituo) [reply spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java:8]: added

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 2 above)
Comment (kaituo) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:150]: if it is null, we will use default result index as before.

File: datasources/src/main/resources/datasources-index-mapping.yml (hunk 3 above)
Comment (kaituo) [reply datasources/src/main/resources/datasources-index-mapping.yml:18]: yeah, we need to save it and retrieve it later.

Comment (penghuo) [reply datasources/src/main/resources/datasources-index-mapping.yml:18]: keyword looks good, no full text search requuired.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 4]
```diff
@@ -63,12 +68,33 @@ public JSONObject getQueryResponse(AsyncQueryJobMetadata asyncQueryJobMetadata)
...
+            ? jobExecutionResponseReader.getResultFromOpensearchIndex(
+                asyncQueryJobMetadata.getJobId(), asyncQueryJobMetadata.getResultIndex())
+            : new JSONObject();
+
+    // if result index document has a status, we are gonna use the status directly; otherwise, we
+    // will use emr-s job status
+    // a job is successful does not mean there is no error in execution. For example, even if result
+    // index mapping
+    // is incorrect, we still write query result and let the job finish.
+    if (result.has(DATA_FIELD)) {
```
Comment (vamsi-amazon) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:83]: We are setting result object only when job status from EMR is success. 

In case of index query, EMR JobRUn State is RUNNING, but the status should be succesful in the result.

File: spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java [hunk 5]
```diff
@@ -8,13 +8,22 @@
...
+
+  // EMR-S will download JAR to local maven
   public static final String SPARK_SQL_APPLICATION_JAR =
-      "s3://flint-data-dp-eu-west-1-beta/code/flint/sql-job.jar";
+      "file:///home/hadoop/.ivy2/jars/org.opensearch_opensearch-spark-sql-application_2.12-0.1.0-SNAPSHOT.jar";
   public static final String SPARK_RESPONSE_BUFFER_INDEX_NAME = ".query_execution_result";
   // TODO should be replaced with mvn jar.
   public static final String FLINT_INTEGRATION_JAR =
-      "s3://spark-datasource/flint-spark-integration-assembly-0.1.0-SNAPSHOT.jar";
+      "s3://flint-data-dp-eu-west-1-beta/code/flint/sql-job.jar";
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java:26]: revert this, it is for Spark datasource, not used in flint.

**Review by penghuo — DISMISSED** [review]

File: datasources/src/main/resources/datasources-index-mapping.yml (hunk 3 above)
Comment (vamsi-amazon) [reply datasources/src/main/resources/datasources-index-mapping.yml:18]: I have skipped properties because we are never going to query based on it. There is no harms in having it...in case some use case comes up.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 4 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:83]: Probably we can merge and handle this use case as bug separately in a separate PR.

**Review by vamsi-amazon — DISMISSED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java (hunk 5 above)
Comment (kaituo) [reply spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java]: reverted

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 4 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:83]: let's track in issue https://github.com/opensearch-project/sql/issues/2214

**Review by penghuo — APPROVED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 4 above)
Comment (kaituo) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:83]: yeah, we can handle this use case a bug separately.

**Review by vamsi-amazon — APPROVED** [review]
